File: 0111-minimum-depth-of-binary-tree/0111-minimum-depth-of-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def minDepth(self, root: Optional[TreeNode]) -> int:
        def dfs(root):
            if root == None:
                return 0
            if root.left == None:
                return 1 + dfs(root.right)
            if root.right == None:
                return 1 + dfs(root.left)
            if root.left != None and root.right != None:
                return 1 + min(dfs(root.left), dfs(root.right))
        return dfs(root)

File: 0111-minimum-depth-of-binary-tree/0111-minimum-depth-of-binary-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int minDepth(TreeNode* root) {
        return dfs(root);
    }
    int dfs(TreeNode* root) {
        if (root == nullptr) {
            return 0;
        }
        if (root->left == nullptr) {
            return 1 + dfs(root->right);
        }
        if (root->right == nullptr) {
            return 1 + dfs(root->left);
        }
        if (root->left != nullptr && root->right != nullptr) {
            return 1 + min(dfs(root->left), dfs(root->right));
        }
        return 0;
    }
};